File: YounSangHo/Week2/2023-09-24_linked_list_cycle_II.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var detectCycle = function (head) {
  if (!head) {
    return null;
  }

  // Phase 1: Determine if a cycle exists
  let slow = head;
  let fast = head;

  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;

    if (slow === fast) {
      break; // Cycle detected
    }
  }

  if (!fast || !fast.next) {
    return null; // No cycle
  }

  // Phase 2: Find the start of the cycle
  slow = head;
  while (slow !== fast) {
    slow = slow.next;
    fast = fast.next;
  }

  return slow;
};

File: SangmunKim/2023-09-24-Linked-List-Cycle-II.py
class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        rabbit = head
        turtle = head

        while rabbit and rabbit.next:
            turtle = turtle.next
            rabbit = rabbit.next.next

            if turtle == rabbit:
                break
        
        if not rabbit or not rabbit.next:
            return None
        
        rabbit = head

        while turtle != rabbit:
            turtle = turtle.next
            rabbit = rabbit.next
        
        return turtle
